src/core/net/ip6.cpp

```diff
@@ -145,61 +145,98 @@ ThreadError Ip6::AddMplOption(Message &message, Header &header, IpProto nextHead
 
     SuccessOrExit(error = message.Prepend(&mplOption, mplOption.GetTotalLength()));
     SuccessOrExit(error = message.Prepend(&hbhHeader, sizeof(hbhHeader)));
-    header.SetPayloadLength(sizeof(hbhHeader) + sizeof(mplOption) + payloadLength);
+    header.SetPayloadLength(header.GetPayloadLength() + sizeof(hbhHeader) + sizeof(mplOption));
     header.SetNextHeader(kProtoHopOpts);
 exit:
     return error;
 }
 
-ThreadError Ip6::InsertMplOption(Message &aMessage, Header &aIp6Header)
+ThreadError Ip6::AddTunneledMplOption(Message &aMessage, Header &aHeader, MessageInfo &aMessageInfo)
 {
     ThreadError error = kThreadError_None;
+    Header tunnelHeader;
+    Address tunnelAddress;
+    const NetifUnicastAddress *source;
+
+    // Use IP-in-IP encapsulation (RFC2473) and ALL_MPL_FORWARDERS address.
+    memset(&tunnelAddress, 0, sizeof(tunnelAddress));
+    tunnelAddress.mFields.m16[0] = HostSwap16(0xff03);
+    tunnelAddress.mFields.m16[7] = HostSwap16(0x00fc);
+
+    tunnelHeader.Init();
+    tunnelHeader.SetHopLimit(static_cast<uint8_t>(kDefaultHopLimit));
+    tunnelHeader.SetPayloadLength(aHeader.GetPayloadLength() + sizeof(tunnelHeader));
+    tunnelHeader.SetDestination(tunnelAddress);
+    tunnelHeader.SetNextHeader(kProtoIp6);
+
+    VerifyOrExit((source = SelectSourceAddress(tunnelAddress, aMessageInfo)) != NULL,
```

Comment:
It might be cleaner to make a copy of `aMessageInfo` with the new peer address value.  This would avoid needing to add the `aDestination` argument to `SelectSourceAddress()`.